src/raft/tracker.nim

```diff
@@ -31,67 +36,195 @@ type
+func init*(T: type RaftVotes, config: RaftConfig): T =
+  let allNodes = config.currentSet & config.previousSet
+  var r = T(voters: allNodes, current: RaftElectionTracker.init(config.currentSet))
+  if config.isJoint:
+    r.previous = some(RaftElectionTracker.init(config.previousSet))
   return r
 
+proc `=copy`*(d: var RaftVotes; src: RaftVotes) {.error.} =
+ discard
+
 func registerVote*(rv: var RaftVotes, nodeId: RaftNodeId, granted: bool): bool =
-  # TODO: Add support for configuration
-  return rv.current.registerVote(nodeId, granted)
+  var success = rv.current.registerVote(nodeId, granted)
+  if rv.previous.isSome:
+    success = success or rv.previous.get().registerVote(nodeId, granted)
+  return success
 
 func tallyVote*(rv: var RaftVotes): RaftElectionResult =
   # TODO: Add support for configuration
+  if rv.previous.isSome:
+    var electionResult = rv.previous.get.tallyVote()
+    if electionResult != RaftElectionResult.Won:
+      return electionResult
   return rv.current.tallyVote()
 
-func find*(ls: RaftTracker, id: RaftnodeId): Option[RaftFollowerProgressTracker] =
+func contains*(rv: var RaftVotes, id: RaftNodeId): bool =
+  if rv.current.contains(id):
+    return true
+  return rv.previous.isSome and rv.previous.get.contains(id)
+
+func find*(ls: RaftTracker, id: RaftNodeId): Option[RaftFollowerProgressTrackerRef] =
   for follower in ls.progress:
     if follower.id == id:
       return some(follower)
-  return none(RaftFollowerProgressTracker)
+  return none(RaftFollowerProgressTrackerRef)
+
+func new*(T: type RaftFollowerProgressTrackerRef, follower: RaftNodeId, nextIndex: RaftLogIndex, now: times.DateTime): T =
+  T(id: follower, nextIndex: nextIndex, matchIndex: 0, commitIndex: 0, replayedIndex: 0, lastMessageAt: now)
+
+func new*(T: type RaftFollowerProgressTrackerRef, follower: RaftNodeId, nextIndex: RaftLogIndex): T =
+  T(id: follower, nextIndex: nextIndex, matchIndex: 0, commitIndex: 0, replayedIndex: 0)
 
-func initFollowerProgressTracker*(follower: RaftNodeId, nextIndex: RaftLogIndex, now: times.DateTime): RaftFollowerProgressTracker =
-  return RaftFollowerProgressTracker(id: follower, nextIndex: nextIndex, matchIndex: 0, commitIndex: 0, replayedIndex: 0, lastMessageAt: now)
+func find(s: var RaftFollowerProgress, what: RaftNodeId): int =
+  result = -1
+  for i, x in s:
+    if x.id == what:
+      return i
+  return -1
 
-func initTracker*(config: RaftConfig, nextIndex: RaftLogIndex, now: times.DateTime): RaftTracker =
-  var tracker = RaftTracker()
+func setConfig*(tracker: var RaftTracker, config: RaftConfig, nextIndex: RaftLogIndex, now: times.DateTime) =
+
+  tracker.current = @[]
+  tracker.previous = @[]
+  
+  var oldProgress = tracker.progress
+  tracker.progress = @[]
+
+  for s in config.currentSet:
+    # TODO: Add can_vote prop
+    tracker.current.add(s)      
+    let oldp = oldProgress.find(s)
+    if oldp != -1:
+        tracker.progress.add(oldProgress[oldp])
```

Comment:
Too many issues with indentation. You can try running https://github.com/arnetheduck/nph over the entire repo.